languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py

```diff
@@ -51,62 +52,100 @@ def _before_compile(query):
     return _authorize_query(query)
 
 
+Permissions = Optional[Dict[Type[Any], Any]]
+
+
 def _authorize_query(query: Query) -> Optional[Query]:
-    """Authorize an existing query with an oso instance, user and action."""
+    """Authorize an existing query with an Oso instance, user, and checked
+    permissions."""
     # Get the query session.
     session = query.session
 
-    # Check whether this is an oso session.
+    # Check whether this is an Oso session.
     if not isinstance(session, AuthorizedSessionBase):
         # Not an authorized session.
         return None
 
-    oso = session.oso_context["oso"]
+    oso: Oso = session.oso_context["oso"]
     user = session.oso_context["user"]
-    action = session.oso_context["action"]
+    checked_permissions: Permissions = session.oso_context["checked_permissions"]
+
+    # Early return if no authorization is to be applied.
+    if checked_permissions is None:
+        return None
 
     # TODO (dhatch): This is necessary to allow ``authorize_query`` to work
     # on queries that have already been made.  If a query has a LIMIT or OFFSET
     # applied, SQLAlchemy will by default throw an error if filters are applied.
     # This prevents these errors from occuring, but could result in some
     # incorrect queries. We should remove this if possible.
-    query = query.enable_assertions(False)
+    query = query.enable_assertions(False)  # type: ignore
 
     entities = {column["entity"] for column in query.column_descriptions}
     for entity in entities:
         # Only apply authorization to columns that represent a mapper entity.
         if entity is None:
             continue
 
-        authorized_filter = authorize_model(oso, user, action, query.session, entity)
+        # If entity is an alias, retrieve the underlying class.
+        alias = inspect(entity).class_ if isinstance(entity, AliasedClass) else None
+
+        # Only apply authorization to columns that have been specified as
+        # requiring authorization.
+        if alias in checked_permissions:
+            action = checked_permissions[alias]  # type: ignore
+        elif entity in checked_permissions:
+            action = checked_permissions[entity]  # type: ignore
+        else:
+            continue
+
+        session = query.session
+        assert isinstance(session, Session)
```

Comment:
This line is from trying to convince Pyright that `session` is indeed a `Session`. I can remove it if anyone minds.

Reply:
Keep it it's a good idea